plotly/tools.py

```diff
@@ -2871,3 +2891,149 @@ def get_candle_decrease(self):
         return (decrease_x, decrease_close, decrease_dif,
                 stick_decrease_y, stick_decrease_x)
 
+class _Dendrogram(FigureFactory):
+    ''' Returns a Dendrogram figure object
+        Example usage:
+        D = Dendrogram( Z )
+        fig = { 'data':D.data, 'layout':D.layout }
+        py.iplot( fig, filename='Dendro', validate=False )'''
+    
+    def __init__(self, X, orientation='bottom', labels=None, colorscale=None, \
+                 width=700, height=700, xaxis='xaxis', yaxis='yaxis' ):
+        ''' Draw a 2d dendrogram tree 
+            X: Heatmap matrix as array of arrays
+            orientation: 'top', 'right', 'bottom', or 'left'
+            labels: List of axis category labels
+            colorscale: Optional colorscale for dendgrogram tree clusters
+            Returns a dendrogram Plotly figure object '''
+    
+        self.orientation = orientation
+        self.labels = labels
+        self.xaxis = xaxis
+        self.yaxis = yaxis
+        self.data = []
+        self.leaves = []
+        self.sign = { self.xaxis:1, self.yaxis:1 }        
+        self.layout = { self.xaxis:{}, self.yaxis:{} }
+        
+        self.sign[self.xaxis] = 1 if self.orientation in ['left','bottom'] else -1
+        self.sign[self.yaxis] = 1 if self.orientation in ['right','bottom'] else -1        
+        
+        dd_traces, xvals, yvals, ordered_labels, leaves = self.get_dendrogram_traces( X, colorscale )
+        
+        self.labels = ordered_labels
+        self.leaves = leaves
+        yvals_flat = yvals.flatten()
+        xvals_flat = xvals.flatten()
+
+        self.zero_vals = []
+
+        for i in range(len(yvals_flat)):
+            if yvals_flat[i] == 0.0 and xvals_flat[i] not in self.zero_vals:
+                self.zero_vals.append( xvals_flat[i] )
+
+        self.zero_vals.sort() 
+                
+        self.layout = self.set_figure_layout( width, height )
+        self.data = Data( dd_traces )
+
+    def get_color_dict( self, colorscale ):
+        ''' Return colorscale used for dendrogram tree clusters '''
+        
+        # These are the color codes returned for dendrograms
+        # We're replacing them with nicer colors
+        default_colors = {'r':'red','g':'green','b':'blue','c':'cyan',\
+                  'm':'magenta','y':'yellow','k':'black','w':'white'}
+        
+        if colorscale is None:
+            colorscale = [
+                "rgb(0,116,217)",
+                "rgb(255,65,54)",
+                "rgb(133,20,75)",
+                "rgb(255,133,27)",
+                "rgb(255,220,0)",
+                "rgb(61,153,112)"]
+        for i in range(len(default_colors.keys())):
+            k = default_colors.keys()[i] 
+            if i < len( colorscale ):
+                default_colors[k] = colorscale[i] 
```

Comment:
nice :) I might be missing something here, but since `default_colors` is a dict, we don't know the order of `default_colors.keys()`. Instead, an [OrderedDict](https://docs.python.org/2/library/collections.html#collections.OrderedDict) might be a better solution, or at least `sorted(default_colors.keys())`